jena-core/src/main/java/org/apache/jena/mem/GraphMemUsingHashMap.java

```diff
@@ -0,0 +1,774 @@
+        final Node pm = triplePattern.getPredicate();
+        final Node om = triplePattern.getObject();
+        final IteratorFiltering filterIterator;
+
+        if (sm.isConcrete()) { // SPO:S??
+            final var triples = this.triplesBySubject.get(sm.getIndexingValue().hashCode());
+            if(triples == null || triples.isEmpty()) {
+                return NiceIterator.emptyIterator();
+            }
+            if(pm.isConcrete()) { // SPO:SP?
+                if(om.isConcrete()) { // SPO:SPO
+                    filterIterator = new IteratorFiltering(triples.iterator(),
+                            t -> om.sameValueAs(t.getObject())
+                                    && pm.equals(t.getPredicate())
+                                    && sm.equals(t.getSubject()));
+                } else { // SPO:SP*
+                    filterIterator = new IteratorFiltering(triples.iterator(),
+                            t -> pm.equals(t.getPredicate())
+                                    && sm.equals(t.getSubject()));
+                }
+            } else { // SPO:S*?
+                if(om.isConcrete()) { // SPO:S*O
+                    filterIterator = new IteratorFiltering(triples.iterator(),
+                            t -> om.sameValueAs(t.getObject())
+                                    && sm.equals(t.getSubject()));
+                } else { // SPO:S**
+                    filterIterator = new IteratorFiltering(triples.iterator(),
+                            t -> sm.equals(t.getSubject()));
+                }
+            }
+        }
+        else if (om.isConcrete()) { // SPO:*?O
+            var triples = this.triplesByObject.get(om.getIndexingValue().hashCode());
+            if(triples == null || triples.isEmpty()) {
+                return NiceIterator.emptyIterator();
+            }
+            if(pm.isConcrete()) { // SPO:*PO
+                filterIterator = new IteratorFiltering(triples.iterator(),
+                        t -> pm.equals(t.getPredicate())
+                                && om.sameValueAs(t.getObject()));
+            } else { // SPO:**O
+                filterIterator = new IteratorFiltering(triples.iterator(),
+                        t -> om.sameValueAs(t.getObject()));
+            }
+        }
+        else if (pm.isConcrete()) { // SPO:*P*
+            var triples = this.triplesByPredicate.get(pm.getIndexingValue().hashCode());
+            if(triples == null || triples.isEmpty()) {
+                return NiceIterator.emptyIterator();
+            }
+            filterIterator = new IteratorFiltering(triples.iterator(),
+                    t -> pm.equals(t.getPredicate()));
+        }
+        else { // SPO:***
+            /*use predicates because they are rare compared to subjects and objects*/
+            return new IteratorWrapperWithRemove(
+                    new ListsOfTriplesIterator(this.triplesByPredicate.values().iterator()), this);
```

Comment:
Why wrap and return here when you're just going to do that on Line 425 anyway?

Reply:
Thank you! I could simplify the code and fixed it: https://github.com/apache/jena/pull/1273/commits/fb232336accc574d35d71bd47659c87d3f25f065